### PR TITLE
remove crossplane provider operator

### PR DESCRIPTION
**Pull request by qpdpQ**
for the issue: https://github.ibm.com/IBMPrivateCloud/roadmap/issues/55208
add a new parameter `crossplaneProviderRemoval` in CommonService CR aligned with `.features.bedrockshim.enabled` 
it is an indicator to perform gracefully uninstallation for provider operator

Signed-off-by: qpdpQ <liyuchen223@gmail.com>

### CODE REVIEW

File: controllers/bootstrap/init.go [hunk 1]
```diff
@@ -371,6 +399,81 @@ func (b *Bootstrap) DeleteCrossplaneAndProviderSubscription(namespace string) er
...
+				uninstallProvider = true
+			} else {
+				uninstallProvider = false
+				break
+			}
+		} else {
+			uninstallProvider = false
+			break
+		}
+	}
```
Comment (Daniel-Fan) [inline controllers/bootstrap/init.go:429]: I think we might need to re-consider the logic here.

If we some CommonService CR's does not say anything about Crossplane, should we just ignore it?
For example,
1. if `.spec.features.bedrockshim` is `nil`, then we may should skip this CR.
2. If `.sepc.features.bedrock.shim.enable: false`, then we should skip `crossplaneProviderRemoval` as well.
3. From my own opinion, we should only consider `uninstallProvider`/`crossplaneProviderRemoval` when `.sepc.features.bedrock.shim.enable: enable`

File: controllers/bootstrap/init.go [hunk 2]
```diff
@@ -196,19 +196,37 @@ func (b *Bootstrap) CrossplaneOperatorProviderOperator(instance *apiv3.CommonSer
...
+		if removeCrossplaneProvider {
+			//delete Crossplane Provider Subscription
+			klog.Info("try to delete CrossplaneProvider")
+			if DeleteErr := b.DeleteCrossplaneProviderSubscription(b.CSData.ControlNs); DeleteErr != nil {
+				return DeleteErr
+			}
+		}
 
+		if !removeCrossplaneProvider {
+			if updateErr := b.CreateorUpdateCFCrossplaneConfigMap("'false'"); updateErr != nil {
```
[the remark this answers is not in the document]
Comment (qpdpQ) [reply controllers/bootstrap/init.go]: yeah

File: controllers/bootstrap/init.go [hunk 3]
```diff
@@ -196,19 +196,37 @@ func (b *Bootstrap) CrossplaneOperatorProviderOperator(instance *apiv3.CommonSer
...
+				return updateErr
+			}
+		}
```
[the remark this answers is not in the document]
Comment (qpdpQ) [reply controllers/bootstrap/init.go]: sure

File: controllers/bootstrap/init.go (hunk 1 above)
Comment (qpdpQ) [reply controllers/bootstrap/init.go]: sure!

File: controllers/bootstrap/init.go [hunk 4]
```diff
@@ -196,50 +196,76 @@ func (b *Bootstrap) CrossplaneOperatorProviderOperator(instance *apiv3.CommonSer
...
+			if !removeCrossplaneProvider {
+				switch b.CSData.CrossplaneProvider {
+				case "odlm":
+					if err := b.installKubernetesProvider(); err != nil {
+						return err
+					}
+				case "ibmcloud":
+					if err := b.installIBMCloudProvider(); err != nil {
+						return err
+					}
```
Comment (Daniel-Fan) [inline controllers/bootstrap/init.go:254]: Wondering that if we could integrate this part into above section on `if removeCrossplaneProvider {} else {}`

File: controllers/bootstrap/init.go [hunk 5]
```diff
@@ -196,50 +196,76 @@ func (b *Bootstrap) CrossplaneOperatorProviderOperator(instance *apiv3.CommonSer
...
-		} else {
-			isLater = false
-		}
-		//isLater value of false means we install, true means we do not install
-		if !isLater {
+
+			b.CSData.CrossplaneProvider = "odlm"
+			if b.SaasEnable {
+				b.CSData.CrossplaneProvider = "ibmcloud"
+			}
```
Comment (Daniel-Fan) [inline controllers/bootstrap/init.go:238]: This part is only about `Provider`, could we integrate them under `if removeCrossplaneProvider {} else {}`

File: controllers/bootstrap/init.go [hunk 6]
```diff
@@ -196,50 +196,76 @@ func (b *Bootstrap) CrossplaneOperatorProviderOperator(instance *apiv3.CommonSer
...
 		} else {
-			klog.Infof("Crossplane operator already exists at a later version in control namespace. Skipping.")
+			// delete crossplane and provider operator if exist
+			if err := b.DeleteCrossplaneAndProviderSubscription(b.CSData.ControlNs); err != nil {
+				return err
+			}
+			value := "'" + strconv.FormatBool(removeCrossplaneProvider) + "'"
+			if updateErr := b.CreateorUpdateCFCrossplaneConfigMap(value); updateErr != nil {
+				return updateErr
+			}
```
Comment (Daniel-Fan) [inline controllers/bootstrap/init.go:265]: Wondering if we should move this part into `DeleteCrossplaneAndProviderSubscription(b.CSData.ControlNs)`, like what we did for `DeleteCrossplaneProviderSubscription`

I am considering that we should go over all CommonService CRs in the cluster and determine which value should be put into ConfigMap.

Comment (qpdpQ) [reply controllers/bootstrap/init.go]: Yes, we should move it to DeleteCrossplaneAndProviderSubscription

File: controllers/bootstrap/init.go (hunk 5 above)
Comment (qpdpQ) [reply controllers/bootstrap/init.go]: yeah good idea!

File: controllers/bootstrap/init.go (hunk 4 above)
Comment (qpdpQ) [reply controllers/bootstrap/init.go]: yeah good idea

**Review by Daniel-Fan — APPROVED** [review]
/lgtm

**Pull request merged**